virtual_rainforest/core/registry.py

```diff
@@ -0,0 +1,222 @@
+    #        name
+    model_name, model_obj = model[0]
+    if module_name_short != model_obj.model_name:
+        excep = RuntimeError(
+            f"The model_name attribute does not match the module name in {module_name}"
+        )
+        LOGGER.critical(excep)
+        raise excep
+
+    # Register the resulting single model class
+    MODULE_REGISTRY[module_name_short]["model"] = model_obj
+    LOGGER.info(f"Registered class for {module_name_short} model: {model_name}")
+
+
+def register_schema(module_name_short: str, schema_file_path: Path) -> None:
+    """Simple function to add configuration schema to the registry.
+
+    Args:
+        module_name: The name to register the schema under
+        schema_file_path: The file path to the JSON Schema file for the model
+
+    Raises:
+        ValueError: If the module name has already been used to register a schema
+    """
+
+    if (
+        module_name_short in MODULE_REGISTRY
+        and "schema" in MODULE_REGISTRY[module_name_short]
+    ):
+        excep = ValueError(
+            f"The module schema for {module_name_short} is already registered"
+        )
+        LOGGER.critical(excep)
+        raise excep
+
+    try:
+        MODULE_REGISTRY[module_name_short]["schema"] = load_schema(
+            module_name_short, schema_file_path
+        )
+    except Exception as excep:
+        LOGGER.critical(
+            f"Schema registration for {module_name_short} failed: check log"
+        )
+        raise excep
+
+    LOGGER.info(
+        "Schema registered for module %s: %s ", module_name_short, schema_file_path
+    )
+
+
+def register_constants_classes(module: ModuleType, module_name_short: str) -> None:
+    """Simple function to add a constants class to the registry.
+
+    Args:
+        module: A module object that might contain constants classes.
+        module_name_short: The short name of the module
+
+    Raises:
+        ValueError: If the model and class name have already been used to register a
+            constants class
+    """
+
+    # Create the dictionary for constants for this module
+    if "constants" not in MODULE_REGISTRY[module_name_short]:
+        MODULE_REGISTRY[module_name_short]["constants"] = {}
+
+    # Try and get the constants submodule and simply leave if none is present
+    constants_submodule = getattr(module, "constants", None)
+
+    if constants_submodule is None:
+        return
+
+    # Otherwise get any constant dataclasses and register them
+    constants_classes = (
+        class_ for class_ in getmembers(constants_submodule) if is_dataclass(class_[1])
+    )
+
+    for class_name, class_obj in constants_classes:
+        if class_name in MODULE_REGISTRY[module_name_short]["constants"]:
+            excep = ValueError(
+                f"Constants class {module_name_short}.{class_name} already registered"
+            )
+            LOGGER.critical(excep)
+            raise excep
```

Comment:
So, just to confirm I get this right: a module can only have one model (with a model_name equal to the short module name) but can have any number of dataclasses defining constants, correct?

Reply:
I thought I have mentioned this somewhere, but you can simply do `model_name = Path(__file__).stem` and you automatically get the model name equal to the module name without telling the user what to do neither having to check if they are the same, as it is done somewhere else.

Reply:
Actually, we could do this in `BaseModel.__init_subclass`:

```python
setattr(cls, "model_name", cls.__module__.split(".")[2])
```

That takes, for example `virtual_rainforest.models.plants.plants_model` and enforces the class model_name without needing to add it to each model subclass.

Reply:
I'm not sure if this will be picked by `mypy`, though...

Reply:
I'll see if it works, but probably not in this PR!